libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/extensions/coroutines/MapboxNavigation.kt

```diff
@@ -0,0 +1,48 @@
+package com.mapbox.navigation.dropin.extensions.coroutines
+
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.api.directions.v5.models.RouteOptions
+import com.mapbox.navigation.base.route.RouterCallback
+import com.mapbox.navigation.base.route.RouterFailure
+import com.mapbox.navigation.base.route.RouterOrigin
+import com.mapbox.navigation.core.MapboxNavigation
+import kotlinx.coroutines.suspendCancellableCoroutine
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+
+data class RoutesWithOrigin(
+    val routes: List<DirectionsRoute>,
+    val routerOrigin: RouterOrigin
+)
+
+class RequestRoutesError(
+    val reasons: List<RouterFailure>,
+    message: String
+) : Error(message)
+
+suspend fun MapboxNavigation.requestRoutes(routeOptions: RouteOptions): RoutesWithOrigin =
+    suspendCancellableCoroutine { continuation ->
+        requestRoutes(
+            routeOptions,
+            object : RouterCallback {
+                override fun onRoutesReady(
+                    routes: List<DirectionsRoute>,
+                    routerOrigin: RouterOrigin
+                ) {
+                    continuation.resume(RoutesWithOrigin(routes, routerOrigin))
+                }
+
+                override fun onFailure(reasons: List<RouterFailure>, routeOptions: RouteOptions) {
+                    val error = RequestRoutesError(
+                        reasons,
+                        "Route request failed with: $reasons"
+                    )
+                    continuation.resumeWithException(error)
```

Comment:
Why use a throwable to control execution flow instead of something like `Result`? The latter is explicit and won't allow for a mistake in the handling code.

Reply:
This extension method only converts callback API to a suspendible function. Not returning Result gives more flexibility over how exceptions are handled at the call site:

```kotlin
// use try catch
try {
  val v = navigation.requestRoutes(routeOptions)
} catch (e: RequestRoutesError) {
  // handle error here
}

// wrap it in a Result instance
val result: Result<RoutesWithOrigin> = runCatching {
    navigation.requestRoutes(routeOptions)
}
```

Reply:
@LukasPaczos I understand your concern. But, these extensions are not meant to be consumed by the general public.  These are internal, dropin module, low-level/data source APIs that are meant to be called by a domain layer [UseCases that already handle exceptions](https://github.com/mapbox/mapbox-navigation-android/blob/5b25af64d74897fe67e2886828dd5fe26f10414e/libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/usecase/UseCase.kt#L46).

Reply:
There's no advantage to throwing exceptions. This and similar functions should return an Expected<Error, Type> like the majority of the API's in the SDK. The advantage is they are deterministic when exceptions aren't thrown.

Reply:
Since we are not going with coroutine based approach and want to stick to callbacks, I am removing this extension. This is no longer valid.